### ISSUE TITLE
Calculated capillary pressures are wrong for parallel Norne run

### ISSUE THREAD

**Issue by blattms** [OPM/opm-autodiff#470]
When running Norne in parallel it turns out that the capillary pressures are wrong. As I checked all the input values to `BlackoilPropsAdFromDeck::capPress` and all are correct, it seems to me that the problem must appear in the calculations done in opm-matrial (`MaterialLaw::capillaryPressures`).

As I am not at all familiar with this code, can someone tell which capillaryPressures function will get called for Norne. Various models seem to be implemented in opm-material, but i have troubles figuring out which one gets called.

As we only compute on one part of the grid in a parallel run, the same behaviour might also appear for the sequential calculations using only a part of the grid.

**Comment by blattms** [comment]
Ok it gets initialized by
```
auto materialLawManager = std::make_shared<MaterialLawManager>();
materialLawManager->initFromDeck(deck, eclipseState, compressedToCartesianIdx);
```
which really poses a problem. It means that it is not valid in a parallel run whenever it uses the compressedToCartesianIdx. Therefore we would need to communicate the stored information of the MaterialLawManager without knowing its implementation. This will not work if the different implementations
store different kinds of data. Is this the case or is there a common parent class that takes care of the data storage?

**Comment by blattms** [comment]
Can I see from the deck file which laws/model will be used?

**Comment by atgeirr** [comment]
The best person to answer this would be @andlaus but I think he is travelling this week. My best tip is to use the debugger to step through the (quite complicated) code to get a grip on it, I do not have the answers to your questions straightaway.

**Comment by blattms** [comment]
The first problem was that nobody took case of making sure that MaterialLawManager was adapted to the loadbalanced grid. In a parallel the manager for the global grid was used. I changed this locally now.

Still it turns out the values that the values stay incorrect. The problem is that `EclEpsScalingPoints` calculates some extrema (max/min) from local tables. In a parallel run these are just a local extrema. See e.g.
https://github.com/OPM/opm-material/blob/6a6d1d2088/opm/material/fluidmatrixinteractions/EclEpsScalingPoints.hpp#L316 and many others.

For a Norne run this results in the following behaviour for global indes 2368:
sequential run:
```gdb
(gdb) p params.drainageParams()
$5 = (const Opm::EclHysteresisTwoPhaseLawParams<Opm::EclEpsTwoPhaseLaw<Opm::PiecewiseLinearTwoPhaseMaterial<Opm::TwoPhaseMaterialTraits<double, 0, 1>, Opm::PiecewiseLinearTwoPhaseMaterialParams<Opm::TwoPhaseMaterialTraits<double, 0, 1> > >, Opm::EclEpsTwoPhaseLawParams<Opm::PiecewiseLinearTwoPhaseMaterial<Opm::TwoPhaseMaterialTraits<double, 0, 1>, Opm::PiecewiseLinearTwoPhaseMaterialParams<Opm::TwoPhaseMaterialTraits<double, 0, 1> > > > > >::EffLawParams &) @0xe026c78: {
  finalized_ = true, 
  effectiveLawParams_ = std::shared_ptr (count 44431, weak 0) 0x7b65ce0, 
  config_ = std::shared_ptr (count 88862, weak 0) 0x76b7860, 
  unscaledPoints_ = std::shared_ptr (count 44431, weak 0) 0x7b65ad0, 
  scaledPoints_ = {maxPcnw_ = 773592.23022346164, maxKrw_ = 0.5, maxKrn_ = 1, 
    saturationPcPoints_ = {_M_elems = {0.040000000000000001, 1}}, 
    saturationKrwPoints_ = {_M_elems = {0.12, 0.87, 1}}, 
    saturationKrnPoints_ = {_M_elems = {0.040000000000000001, 0.12, 0.87}}}}
```
versus parallel with 2 processes:
```
(gdb) p params.drainageParams()
$9 = (const Opm::EclHysteresisTwoPhaseLawParams<Opm::EclEpsTwoPhaseLaw<Opm::PiecewiseLinearTwoPhaseMaterial<Opm::TwoPhaseMaterialTraits<double, 0, 1>, Opm::PiecewiseLinearTwoPhaseMaterialParams<Opm::TwoPhaseMaterialTraits<double, 0, 1> > >, Opm::EclEpsTwoPhaseLawParams<Opm::PiecewiseLinearTwoPhaseMaterial<Opm::TwoPhaseMaterialTraits<double, 0, 1>, Opm::PiecewiseLinearTwoPhaseMaterialParams<Opm::TwoPhaseMaterialTraits<double, 0, 1> > > > > >::EffLawParams &) @0xdf8f488: {
  finalized_ = true, 
  effectiveLawParams_ = std::shared_ptr (count 44431, weak 0) 0x7b7ec30, 
  config_ = std::shared_ptr (count 88862, weak 0) 0x76d07b0, 
  unscaledPoints_ = std::shared_ptr (count 44431, weak 0) 0x7b7ea20, 
  scaledPoints_ = {maxPcnw_ = 2642291.6690998324, maxKrw_ = 0.5, maxKrn_ = 1, 
    saturationPcPoints_ = {_M_elems = {0.040000000000000001, 1}}, 
    saturationKrwPoints_ = {_M_elems = {0.12, 0.87, 1}}, 
    saturationKrnPoints_ = {_M_elems = {0.040000000000000001, 0.12, 0.87}}}}
```
i need to figure out how to prevent this, but this seems like a major rewrite.

**Comment by blattms** [comment]
I guess I need to do some more digging as the tables are extracted directly from the deck and therefore global. Maybe extractGridProperty is to blame when overwriting the values per cell. We will see...

**Comment by blattms** [comment]
@andlaus Do you have a hint where and how `params.drainageParams().scaledPoints_.maxPcng_` is actually calculated?

**Comment by andlaus** [comment]
that's done in the `EclEpsScalingPoints::init()`. if endpoint scaling is enabled for capillary pressure, the source value is determined by the `EclEpsScalingInfo::extractScaled()` but it gets modified if SWATINIT is also used (which is the case for Norne, IIRC).

**Comment by blattms** [comment]
Thanks. Where and when does it get modified if SWATINIT is present?

**Comment by andlaus** [comment]
in the material law manager. the incriminating method is called by `SaturationPropsFromDeck`.

**Comment by blattms** [comment]
I guess that is done during `Opm::SaturationPropsFromDeck::swatInitScaling` (gdb rules). These implicit interferences between the different property managers are really hard to grasp.

**Comment by blattms** [comment]
Thanks. that might finally be the right track to search further.

**Comment by andlaus** [comment]
well, the `SaturationPropsFromDeck` is just kept as a relatively shallow impedance layer between the opm-material code and the historical API expected by the simulator. removing `SaturationPropsFromDeck` would be beneficial from the code-quality perspective IMO, but it would not yield any "sticker features" and it would be quite a bit of work as far as I can see...

**Comment by atgeirr** [comment]
I assume this issue is still unresolved, but will not treat it as a show-stopper for the upcoming release.

**Comment by blattms** [comment]
Yes, the capillary pressures are still wrong for the full norne in a parallel run. Off I go to debugging mode.

**Comment by blattms** [comment]
No, I did not respect that there is a pressure for each phase and was therefore comparing apples with oranges.
This was resolved with the merge of #523
